python/ray/data/datasource/image_folder_datasource.py

```diff
@@ -0,0 +1,135 @@
+import pathlib
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional, Union
+
+import imageio as iio
+import numpy as np
+from ray.data.block import Block
+from ray.data.datasource.binary_datasource import BinaryDatasource
+from ray.data.datasource.datasource import ReadTask
+from ray.data.datasource.file_based_datasource import _resolve_paths_and_filesystem
+from ray.data.datasource.file_meta_provider import (
+    BaseFileMetadataProvider,
+    DefaultFileMetadataProvider,
+    FastFileMetadataProvider,
+)
+from ray.data.datasource.partitioning import PathPartitionFilter
+
+if TYPE_CHECKING:
+    import pyarrow
+
+IMAGE_EXTENSIONS = [".png", ".jpg", ".jpeg", ".tiff", ".bmp", ".gif"]
+
+
+class ImageFolderDatasource(BinaryDatasource):
+    """A datasource that lets you read datasets like `ImageNet <https://www.image-net.org/>`_.
+
+    This datasource works with any dataset where images are arranged in this way:
+
+    .. code-block::
+
+        root/dog/xxx.png
+        root/dog/xxy.png
+        root/dog/[...]/xxz.png
+
+        root/cat/123.png
+        root/cat/nsdf3.png
+        root/cat/[...]/asd932_.png
+
+    Datasets read with ``ImageFolderDatasource`` contain two columns: ``'image'`` and
+    ``'label'``. The ``'image'`` column contains ``ndarray`` objects of shape 
+    :math:`(H, W, C)`, and the ``label`` column contains strings corresponding to
+    labels.
+
+    Examples:
+        >>> import ray
+        >>> from ray.data.datasource import ImageFolderDatasource
+        >>>
+        >>> ds = ray.data.read_datasource(  # doctest: +SKIP
+        ...     ImageFolderDatasource(),
+        ...     paths=["/data/imagenet/train"]
+        ... )
+        >>> sample = ds.take(1)[0]  # doctest: +SKIP
+        >>> sample["image"].shape  # doctest: +SKIP
+        (469, 387, 3)
+        >>> sample["label"]  # doctest: +SKIP
+        'n01443537'
+
+    Raises:
+        ValueError: if more than one path is provided. You should only provide the path
+            to the dataset root.
+    """
+
+    def prepare_read(
+        self,
+        parallelism: int,
+        paths: Union[str, List[str]],
+        filesystem: Optional["pyarrow.fs.FileSystem"] = None,
+        schema: Optional[Union[type, "pyarrow.lib.Schema"]] = None,
+        open_stream_args: Optional[Dict[str, Any]] = None,
+        meta_provider: BaseFileMetadataProvider = DefaultFileMetadataProvider(),
+        partition_filter: PathPartitionFilter = None,
+        # TODO(ekl) deprecate this once read fusion is available.
+        _block_udf: Optional[Callable[[Block], Block]] = None,
+        **reader_args,
+    ) -> List[ReadTask]:
+        if len(paths) > 1:
```

Comment:
Make sure it's not empty either?
```suggestion
        if len(paths) != 1:
```